File: backend/app/schemas/imports.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.track import TrackResponse


class ImportRootInfo(BaseModel):
    id: str
    label: str


class ImportConfigurationResponse(BaseModel):
    enabled: bool
    message: str
    roots: list[ImportRootInfo]


class ImportResolvedPath(BaseModel):
    root_id: str
    relative_path: str
    display_path: str


class ImportScanRequest(BaseModel):
    root_id: str
    relative_subdir: str | None = None


class ImportScanLimits(BaseModel):
    max_files: int
    max_depth: int
    max_file_size_bytes: int


class ImportScanCandidate(BaseModel):
    relative_path: str
    basename: str
    extension: str
    size_bytes: int
    status: str = "supported"
    media_kind: str = "audio"


class ImportScanSkippedItem(BaseModel):
    relative_path: str
    basename: str
    extension: str | None = None
    size_bytes: int | None = None
    status: str = "skipped"
    reason: str


class ImportScanResponse(BaseModel):
    enabled: bool
    message: str
    root: ImportRootInfo | None = None
    scanned_relative_path: str | None = None
    candidates: list[ImportScanCandidate]
    skipped: list[ImportScanSkippedItem]
    limits: ImportScanLimits


class ImportConfirmFileSelection(BaseModel):
    relative_path: str


class ImportConfirmRequest(BaseModel):
    root_id: str
    files: list[ImportConfirmFileSelection]


class ImportDuplicateWarning(BaseModel):
    match_type: str
    reason: str
    candidate_track_ids: list[int]


class ImportConfirmResult(BaseModel):
    relative_path: str
    basename: str
    status: str
    media_kind: str | None = None
    track: TrackResponse | None = None
    error: str | None = None
    duplicate_warnings: list[ImportDuplicateWarning] = Field(default_factory=list)


class ImportConfirmResponse(BaseModel):
    enabled: bool
    message: str
    root: ImportRootInfo | None = None
    batch_id: int | None = None
    requested_count: int
    imported_count: int
    skipped_count: int
    failed_count: int
    results: list[ImportConfirmResult]


class ImportBatchItemResponse(BaseModel):
    id: int
    relative_path: str
    basename: str
    status: str
    track_id: int | None = None
    track: TrackResponse | None = None
    error: str | None = None
    media_kind: str | None = None
    created_at: datetime
    updated_at: datetime


class ImportBatchResponse(BaseModel):
    id: int
    root: ImportRootInfo
    status: str
    message: str | None = None
    requested_count: int
    imported_count: int
    skipped_count: int
    failed_count: int
    items: list[ImportBatchItemResponse]
    created_at: datetime
    updated_at: datetime

File: web/src/types/imports.ts
import type { Track } from "./track";

export type ImportRootInfo = {
  id: string;
  label: string;
};

export type ImportConfigurationResponse = {
  enabled: boolean;
  message: string;
  roots: ImportRootInfo[];
};

export type ImportScanLimits = {
  max_files: number;
  max_depth: number;
  max_file_size_bytes: number;
};

export type ImportScanCandidate = {
  relative_path: string;
  basename: string;
  extension: string;
  size_bytes: number;
  status: "supported" | string;
  media_kind: "audio" | "video" | string;
};

export type ImportScanSkippedItem = {
  relative_path: string;
  basename: string;
  extension: string | null;
  size_bytes: number | null;
  status: "skipped" | string;
  reason: string;
};

export type ImportScanRequest = {
  root_id: string;
  relative_subdir?: string | null;
};

export type ImportScanResponse = {
  enabled: boolean;
  message: string;
  root: ImportRootInfo | null;
  scanned_relative_path: string | null;
  candidates: ImportScanCandidate[];
  skipped: ImportScanSkippedItem[];
  limits: ImportScanLimits;
};

export type ImportConfirmRequest = {
  root_id: string;
  files: Array<{
    relative_path: string;
  }>;
};

export type ImportDuplicateWarning = {
  match_type: string;
  reason: string;
  candidate_track_ids: number[];
};

export type ImportConfirmResult = {
  relative_path: string;
  basename: string;
  status: "imported" | "skipped" | "failed" | string;
  media_kind: "audio" | "video" | string | null;
  track: Track | null;
  error: string | null;
  duplicate_warnings: ImportDuplicateWarning[];
};

export type ImportConfirmResponse = {
  enabled: boolean;
  message: string;
  root: ImportRootInfo | null;
  batch_id: number | null;
  requested_count: number;
  imported_count: number;
  skipped_count: number;
  failed_count: number;
  results: ImportConfirmResult[];
};

export type ImportBatchItemResponse = {
  id: number;
  relative_path: string;
  basename: string;
  status: "imported" | "skipped" | "failed" | string;
  track_id: number | null;
  track: Track | null;
  error: string | null;
  media_kind: string | null;
  created_at: string;
  updated_at: string;
};

export type ImportBatchResponse = {
  id: number;
  root: ImportRootInfo;
  status: "imported" | "skipped" | "failed" | "importing" | string;
  message: string | null;
  requested_count: number;
  imported_count: number;
  skipped_count: number;
  failed_count: number;
  items: ImportBatchItemResponse[];
  created_at: string;
  updated_at: string;
};
